components/dispute-game-contracts/README.md

```diff
@@ -0,0 +1,276 @@
+2. Define $M_{i+1} = progress(M_i)$
+3. Compute $M_1, M_2, ...$ until you arrive at a "terminal" state $M_t$
+4. Check that $M_t.chainState.root == R'$. If yes, $(R, transactions) \to R'$ is valid; else, it is invalid.
+
+In the event of a validator disputing a sequencer-posted state root, the "bisection game" plays out as follows.
+
+1. [Off-chain] The sequencer computes _its_ belief about what the correct sequence of machine states is, generating a sequence $S_0, ..., S_n$. (A correctly behaving sequencer would find $M_i = S_i$.)
+2. [Off-chain] The validator computes _their_ belief about what the correct sequence of machine states is, generating a sequence $V_0, ..., V_m$. (A correctly behaving validator would find $M_i = V_i$.)
+3. [On-chain] The sequencer commits to the root of the machine state at some steps $step_0, step_1, ..., step_k$, where initially, $step_k = \lfloor k/n \rfloor$.
+4. [On-chain] The validator compares $S_{step_i}.root$ to $V_{step_i}.root$ for each $i$, finding the first $i$ such that they differ. (This implies $S_{step_{i-1}} \to S_{step_i}$ is an invalid transition). They assert on-chain that $step_i$ is invalid, forcing the sequencer to split the transition from $step_{i-1}$ to $step_i$ into k smaller transitions.
+5. Repeat 4, narrowing down the values of $step_i$ until the validator can directly prove an invalid transition on-chain.
+
+**Comment:** This seems different from the game played in the Arbitrum contract. In an Arbitrum challenge, the proposer seems to initially commit to the root of a binary Merkle tree formed from the [entire execution trace](https://github.com/OffchainLabs/arbitrum/blob/41858dc49be854188ec821ed956e767db1e617a9/packages/arb-bridge-eth/contracts/challenge/Challenge.sol#L126). A path down that tree is [revealed by the proposer](https://github.com/OffchainLabs/arbitrum/blob/41858dc49be854188ec821ed956e767db1e617a9/packages/arb-bridge-eth/contracts/challenge/Challenge.sol#L189-L194). Thus, `Challenge.sol` seems to require the sequencer and challenger to hash every machine state up front, while this design allows the sequencer to hash machine states that are committed to on-chain.
+
+Working prototype code can be found in [this repo](src/bisection.ts)), written in Typescript.
+
+## Initializing a challenge
+
+Assume that a `Challenge.sol` contract is used to adjudicate a single challenge. There may be a `ChallengeFactory.sol` contract which deploys new `Challenge` contracts — this is how Arbitrum does it, for instance.
+
+The challenger deploys a challenge contract instance to manage the challenge for the transition T. A commitment to R, R' and `transactions` should be recorded.
+
+This logic is not currently specified.
+
+## First commitment
+
+The sequencer constructs some arbitrary sequence $S_0, S_1, ..., S_n$ and commits to the $k+1$ steps $[(S_0.root, 0), (S_{\lfloor n/k\rfloor}.root, \lfloor n/k \rfloor),..., (S_n.root, n)]$ on-chain:
```

Comment:
And to be clear, this is the **no upfront execution trace approach** - right?

Reply:
that's right!

---

I thought a little more about it, and it's worth pointing out that you can use the same approach with an **upfront commitment to the whole execution trace**:
- The sequencer computes the hashes of the machine states `S_0, ..., S_n`
- They place the hashes as leaves in a balanced, k-ary Merkle tree
- They commit to the first level of the tree (they could commit to the root, but this saves a step)

Then, the verifier points to one of the currently committed to nodes. The rules dictate that the sequencer must then reveal the children of that node. The verifier can find which node to point to by constructing the correct Merkle tree, and comparing the values posted vs the local values.

(If you want to save gas, you can avoid verifying Merkle proofs, since the verifier can verify the Merkle proofs locally. you just need to allow the verifier to point out "your Merkle tree is invalid" to win the game.)

So, whether the game is searching for a leaf in a Merkle tree, or searching for an element of an array, you can implement the game as:
- sequencer posts an array of hashes
- verifier points to one of those hashes

The interpretation of the hashes differs, and the exit logic differs accordingly.

---

In the "explore the Merkle tree of the execution trace" game, it's easier to see how to halve the number of rounds:
- on their turn, each person finds the leftmost node that they contest, and they expand it according to _their_ copy of the Merkle tree
- if Alice reveals the final row of leaves, then Bob proves a proof of an invalid transition to a state corresponding to one of those leaves. (A correct actor can guarantee that if _they_ reveal a row of leaves, that every machine state corresponding to those leaves is valid.)

Reply:
> I thought a little more about it, and it's worth pointing out that you can use the same approach with an **upfront commitment to the whole execution trace**:
> 
> * The sequencer computes the hashes of the machine states `S_0, ..., S_n`
> * They place the hashes as leaves in a balanced, k-ary Merkle tree
> * They commit to the first level of the tree (they could commit to the root, but this saves a step)
> 
> Then, the verifier points to one of the currently committed to nodes. The rules dictate that the sequencer must then reveal the children of that node. The verifier can find which node to point to by constructing the correct Merkle tree, and comparing the values posted vs the local values.
> 
> (If you want to save gas, you can avoid verifying Merkle proofs, since the verifier can verify the Merkle proofs locally. you just need to allow the verifier to point out "your Merkle tree is invalid" to win the game.)


FYI @kerzhner is going to code up this idea to see how it looks.

I just discussed it with him; it seems like it's not too bad, but you have to worry about more edge cases. Plus, you are forced to supply two Merkle proofs, costing `~k log_k(n)` hashes to provide witnesses for a k-ary Merkle tree commitment to `n` machine steps. You could use a binary Merkle tree, skipping k levels at a time, at the cost of additional complexity and edge cases.

As a comparison with the simpler design outlined in this PR:
- ❌ the implementation is more difficult complex
- ❌ single step fraud proofs are more expensive
- ✅ it supports being able to refer to _any_ state prior to the first disputed state, rather than just the step immediately prior to the first disputed state
- ✅ no Merkle proofs are needed in the interactive phase
- 🏳️ (neutral) Since both parties need to compute every machine state, they've already computed all of the hashes of the machine state. The "commit to a Merkle root" approach requires both parties to compute ~twice as many hashes to fill in the tree, but that's negligible compared to creating the execution trace.